network/tls.go

```diff
@@ -0,0 +1,415 @@
+// The CN is always the hex-encoded form of the conode's public key.
+//
+// Because each side needs a nonce which is controlled by the opposite party in the
+// mutual authentication, but TLS does not support sending application data before the handshake,
+// we need to find places in the normal TLS 1.2 handshake where we can "tunnel" the nonce
+// through. On the TLS client side, we use ClientHelloInfo.ServerName. On the
+// the TLS server side, we use the.
+//
+// There is a risk that with a less customizable TLS implementation than
+// Go's, it would not be possible to send the nonces through like this. However,
+// for the moment we are not targeting other languages than Go on the conode/conode
+// communication channel.
+
+// TODO: Websockets.
+// All of this is completely unrelated to HTTPS security on the websocket side. For
+// that, we will implement an opt-in Let's Encrypt client in websocket.go.
+
+// certMaker holds the data necessary to make a certificate on the fly
+// and give it to crypto/tls via the GetCertificate and
+// GetClientCertificate callbacks.
+type certMaker struct {
+	si      *ServerIdentity
+	suite   Suite
+	subj    pkix.Name
+	subjDer []byte
+	k       *ecdsa.PrivateKey
+}
+
+func newCertMaker(s Suite, si *ServerIdentity) (*certMaker, error) {
+	cm := &certMaker{
+		si:    si,
+		suite: s,
+	}
+
+	k, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	cm.k = k
+
+	cm.subj = pkix.Name{CommonName: cm.si.Public.String()}
+	der, err := asn1.Marshal(cm.subj.CommonName)
+	if err != nil {
+		panic("unexpected asn.1 marshal failure")
+	}
+	cm.subjDer = der
+	return cm, nil
+}
+
+func (cm *certMaker) getCertificate(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	return cm.get([]byte(hello.ServerName))
+}
+
+func (cm *certMaker) getClientCertificate(req *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+	if len(req.AcceptableCAs) == 0 {
+		return nil, errors.New("server did not provide a nonce in AcceptableCAs")
+	}
+	return cm.get(req.AcceptableCAs[0])
+}
+
+func (cm *certMaker) get(nonce []byte) (*tls.Certificate, error) {
+	if len(nonce) != nonceSize {
+		return nil, errors.New("nonce is the wrong size")
+	}
+
+	// Create a signature that proves that:
+	// 1. since the nonce was generated by the peer,
+	// 2. for this public key,
+	// 3. we have control of the private key that is associated with the public
+	// key named in the CN.
+	//
+	// Do this using the same standardized ASN.1 marshaling that x509 uses so
+	// that anyone trying to check these signatures themselves in antoher language
+	// will be able to easily do so with their own x509 + kyber implementation.
+	buf := bytes.NewBuffer(nonce)
+	buf.Write(cm.subjDer)
+	sig, err := schnorr.Sign(cm.suite, cm.si.private, buf.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	// The serial number is non-security critical how we are using it here,
+	// so just make it a big random number.
```

Comment:
Is this comment still up to date? It's using the random number generator from kyber, which I imagine is secure.